internal/graph/storagewrapper.go

```diff
@@ -0,0 +1,340 @@
+	)
+	return c.newCachedIterator(ctx, iter, b.String())
+}
+
+// newCachedIterator either returns a cached static iterator for a cache hit, or
+// returns a new iterator that attempts to cache the results.
+func (c *CachedDatastore) newCachedIterator(
+	ctx context.Context,
+	dsIterFunc iterFunc,
+	key string,
+) (storage.TupleIterator, error) {
+	span := trace.SpanFromContext(ctx)
+	tuplesCacheTotalCounter.Inc()
+
+	cacheKey := cacheKeyFor(key)
+
+	if cacheKey == "" {
+		return dsIterFunc(ctx)
+	}
+
+	cachedResp := c.cache.Get(cacheKey)
+	isCached := cachedResp != nil && !cachedResp.Expired && cachedResp.Value != nil
+
+	if isCached {
+		tuplesCacheHitCounter.Inc()
+		span.SetAttributes(attribute.Bool("cached", true))
+		return storage.NewStaticTupleIterator(cachedResp.Value.([]*openfgav1.Tuple)), nil
+	}
+
+	iter, err := dsIterFunc(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	return &cachedIterator{
+		iter:          iter,
+		tuples:        make([]*openfgav1.Tuple, 0, c.maxResultSize),
+		cacheKey:      cacheKey,
+		cache:         c.cache,
+		maxResultSize: c.maxResultSize,
+		ttl:           c.ttl,
+	}, nil
+}
+
+// Close closes the datastore and cleans up any residual resources.
+func (c *CachedDatastore) Close() {
+	c.OpenFGADatastore.Close()
+}
+
+type cachedIterator struct {
+	iter     storage.TupleIterator
+	tuples   []*openfgav1.Tuple
+	cacheKey string
+	cache    storage.InMemoryCache[any]
+	ttl      time.Duration
+
+	// maxResultSize is the maximum number of tuples to cache. If the number
+	// of tuples found exceeds this value, it will not be cached.
+	maxResultSize int
+
+	// sf is used to prevent draining the same iterator
+	// across multiple requests.
+	sf singleflight.Group
+
+	// closeOnce is used to synchronize .Close() and ensure
+	// it's only done once.
+	closeOnce sync.Once
+
+	// wg is used purely for testing and is an internal detail.
+	wg sync.WaitGroup
+}
+
+// Next see [Iterator.Next].
+func (c *cachedIterator) Next(ctx context.Context) (*openfgav1.Tuple, error) {
+	t, err := c.iter.Next(ctx)
+	if err != nil {
+		if !errors.Is(err, storage.ErrIteratorDone) &&
+			!errors.Is(err, context.Canceled) &&
+			!errors.Is(err, context.DeadlineExceeded) {
+			c.tuples = nil // don't store results that are incomplete
+		}
+		return nil, err
+	}
+
+	c.addToBuffer(t)
+
+	return t, nil
+}
+
+// Stop see [Iterator.Stop].
+func (c *cachedIterator) Stop() {
+	c.closeOnce.Do(func() {
+		if c.tuples == nil {
+			c.iter.Stop()
+			return
+		}
+		// prevent goroutine if iterator was already consumed
+		ctx := context.Background()
+		if _, err := c.iter.Head(ctx); errors.Is(err, storage.ErrIteratorDone) {
+			c.flush()
+			c.iter.Stop()
+			return
+		}
+
+		c.wg.Add(1)
+		go func() {
+			defer c.iter.Stop()
+			defer c.wg.Done()
+			// prevent draining on the same iterator across multiple requests
+			_, _, _ = c.sf.Do(c.cacheKey, func() (interface{}, error) {
```

Comment:
i'm confused about why we need the `sf` object if we already had the `closeOnce` object in line 258? is that not enough to guarantee that this whole function is only called once?

Reply:
There may be 2 instances of the iterator. Say A and B, which both are working with the same `cacheKey`. `closeOnce` doesn't prevent A from closing B or vice versa (as intended). But it would be pointless for both A and B to drain the same iterator (based on the assumption of sharing the same cacheKey). So only 1 of them needs to drain and store in cache.

Reply:
wired this up correctly in https://github.com/openfga/openfga/pull/1924/commits/150f7da03e4b981b55efce8997ec3597930931eb